File: eco-core/core-plugin/src/main/java/com/willfp/eco/spigot/eventlisteners/NaturalExpGainBuilder.java
package com.willfp.eco.spigot.eventlisteners;

import com.willfp.eco.core.events.NaturalExpGainEvent;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang.Validate;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.event.player.PlayerExpChangeEvent;
import org.jetbrains.annotations.NotNull;

class NaturalExpGainBuilder {
    @Getter
    @Setter
    private boolean cancelled = false;

    @Getter
    @Setter
    private PlayerExpChangeEvent event;

    @Getter
    @Setter
    private Location location;
    
    @Getter
    @Setter
    private BuildReason reason;
    
    NaturalExpGainBuilder(@NotNull final BuildReason reason) {
        this.reason = reason;
    }

    public void push() {
        Validate.notNull(event);
        if (this.cancelled) {
            return;
        }

        NaturalExpGainEvent naturalExpGainEvent = new NaturalExpGainEvent(event);

        Bukkit.getPluginManager().callEvent(naturalExpGainEvent);
    }

    public enum BuildReason {
        BOTTLE,
        PLAYER
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/eco/internal/spigot/eventlisteners/NaturalExpGainBuilder.kt
package com.willfp.eco.internal.spigot.eventlisteners

import com.willfp.eco.core.events.NaturalExpGainEvent
import moss.factions.shade.com.google.common.base.Preconditions
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.event.player.PlayerExpChangeEvent

internal class NaturalExpGainBuilder(var reason: BuildReason) {
    private var cancelled = false
    var event: PlayerExpChangeEvent? = null
    var location: Location? = null

    fun push() {
        Preconditions.checkNotNull(event, "PlayerExpChangeEvent cannot be null!")
        if (cancelled) {
            return
        }
        val naturalExpGainEvent = NaturalExpGainEvent(event!!)
        Bukkit.getPluginManager().callEvent(naturalExpGainEvent)
    }

    enum class BuildReason {
        BOTTLE, PLAYER
    }
}